### ISSUE TITLE
Fileformat: CAST ( .LST plain text / .CST 8-bit )

### ISSUE THREAD

**Issue by inflex** [OpenBoardView/OpenBoardView#59]
File format to implement, at least this one looks moderately rudimentary being plain text (the .LST parts at least, the CST seems 8-bit)

This package contains a viewer as well, and a pdf

[cast.zip](https://github.com/OpenBoardView/OpenBoardView/files/591816/cast.zip)

**Comment by piernov** [comment]
![capture d ecran_2016-11-15_22-06-21](https://cloud.githubusercontent.com/assets/3323883/20324178/cdd74ed0-ab7f-11e6-8bab-e1e65402f164.png)
Nowhere near being usable but it's a start.

**Comment by inflex** [comment]
Less than 12 hours and you're already at that stage, I'm impressed!

What's left still to do?

**Comment by piernov** [comment]
For the most basic support, I have to figure out which field tells on which side the component is (should not be too hard) and I have to parse the nets list and find the corresponding field in the pins (should be rather easy too). Also the current code is a dirty mess. I still have to find out where the board outline is defined. Then come the test points.
This format also packs in a few extra features not supported in OBV like vias, traces and pin/component shapes.
Like FZ it provides extra information on the parts in the LST file which is not parsed at all for now.

**Comment by inflex** [comment]
A shapes library, with standard outlines is certainly something that needs to be worked on in the future, though at least for now the rendering is "sufficiently ample" for the task at hand, but it's always nice to do better when there's nothing else left to do :+1: 

The part information at least we can currently display like we do with the FZ format.  While it's not separated/parsed in to smaller units (name/value/outline) at least the collective information can still be shown [in the side bar or hover]

**Comment by piernov** [comment]
![capture d ecran_2016-11-16_08-21-10](https://cloud.githubusercontent.com/assets/3323883/20338173/ac60ff16-abd5-11e6-8eec-612a90144633.png)

Looks better. The side of the board is defined by the layer, I need to figure out if there's way to know the number of layers, for now 1 & 12 are hardcoded.

**Comment by inflex** [comment]
Can you process using 1 = top (or bottom), anything > 1 = other side?  Or wait till you know the maximum number of layers and then post-process the data to set the side accordingly?

**Comment by piernov** [comment]
Committed my work here: https://github.com/piernov/OpenBoardView/commit/0b8ce0f394751e2fca0463a991c69b27eee7dd8b
Not finished but I'm not sure I will have the time to do it…

**Comment by inflex** [comment]
Do you think it might be worth pushing that to master?  The testing I did worked fine.

**Comment by lordelectrode** [comment]
Just gonna link relevant fixme for hardcoded layer 12: https://github.com/OpenBoardView/OpenBoardView/blob/master/src/openboardview/FileFormats/CSTFile.cpp#L94

**Comment by piernov** [comment]
I guess @inflex suggestion is relevant, I can't find where the total number of layers would be defined.

**Comment by lordelectrode** [comment]
`if (layer > 0x01)` seems to work

**Comment by inflex** [comment]
I'd be more inclined to set it to > 5 or 6 probably, as it's entirely possible that we may run in to future boardview files that perhaps use 1..4 perhaps as top layers.   It's a guess only but I think it'd be prudent to leave a little room for other potential "top layers".

**Comment by piernov** [comment]
I assume those are only electrical layers. It would be weird to put components on the inner layers…

**Comment by inflex** [comment]
I agree, it would be strange to do that @piernov , but not to say It wouldn't happen.  Maybe they export-mapped to 1 for outline and 2 for top. Not sure, but just thought at least if there was any data in there, it'd still be mapped as a failsafe.  Or we could just dump a warning.
